### ISSUE TITLE
Bug with wildcard / substring / suffix search in mod_vcard

### ISSUE THREAD

**Issue by MatthiasKohl** [processone/ejabberd#633]
Not sure if this is the best way to report issues for ejabberd but here we go:
I'm using ejabberd 15.04.51 with Erlang R16B03 (erts-5.10.4) 64-bit and mnesia as the backend.
I wanted to use the suffix search for vCards using a wildcard (like searching for nick with "Some*" which should return all vCards having a (lowercase) nick starting with "some").

This doesn't work and I looked a bit into the code why. From the history of mod_vcard.erl, I can see that this must have worked right up until a big commit in March 2013 named 'Accumulated patch to binarize and indent code'.
In Erlang ETS tables, you can search for suffixes using the "Suffix" ++ ' _ ' syntax, but this doesn't work with binaries. The make_val/1 function in mod_vcard now just returns [<<"Suffix">>] ++ ' _ ' which doesn't make sense at all.
On top of that, ETS doesn't allow binary pattern matching and the Erlang guards are not sufficient either to check if a binary string is a prefix/suffix of another one.
Do you have any idea how to make this work without going back to non-binary strings or changing the DB ?

**Comment by mremond** [comment]
yes, this is a good way to report the issue. Thanks !

**Comment by MatthiasKohl** [comment]
Cool, so to workaround this, I found 2 solutions.
1. Just convert everything from binaries to strings and back and use strings as the actual data type in mnesia. I really don't like that because I can't do that easily in a custom module without breaking stuff (some admin functionality directly calls mod_vcard, which expects binaries in the table).
2. Iterate over the table (which I'm doing using mnesia:dirty_slots). It's quite inefficient I guess, but not too bad, and at least I can do this in a custom module by turning mod_vcard off.
If you would like me to share the code for any of those solutions, I'm all for it, it's not that complicated, but like I said, I don't think these are the best solutions.
Cheers

**Comment by badlop** [comment]
Right, when searching for joe*, this is called:
```
mnesia:dirty_select(vcard_search,[{{vcard_search,'_','_','_','_',[<<"joe">>,'_'],
  '_','_','_','_','_','_','_','_','_','_','_','_','_','_','_','_','_','_','_','_'},
 [], ['$_']}]).
```

I don't know if there's some way in mnesia to do that search wildcard. I found a first approximation, which isn't perfect because it uses the arithmetic >=
```
mnesia:dirty_select(vcard_search,[{{vcard_search,'_','_','_','_','$1',
 '_','_','_','_','_','_','_','_','_','_','_','_','_','_','_','_','_','_','_','_'},
  [{'>=', '$1', <<"joe">>}], ['$_']}]).
```

**Comment by MatthiasKohl** [comment]
I'm quite sure that you can't do the matching directly with a select query. Binary pattern matching is not implemented (like list pattern matching), see http://erlang.2086793.n4.nabble.com/Binary-pattern-matching-on-ETS-td4708884.html.
And there is no special syntax for matching binary prefixes (at least, not documented).

The approach with the arithmetic >= is probably better than what I found so far (iterating through the table and filtering the values manually), but of course, you need to have two bounds:
Basically, X >= <<"joe">> andalso X < <<"jof">>.
That <<"jof">> is something like <<(binary:part(<<"joe">>, 0, byte_size(<<"joe">>) - 1))/binary, (binary:last(<<"joe">>) + 1)>>.

**Comment by slesru** [comment]
Hello!

Could you tell me is there fix for this issue?

Thank you!

**Comment by badlop** [comment]
The old tricky solution I mentioned, `[{'>=', '$1', <<"joe">>}]`, is unacceptable because it matches "ja", "je"...

I've just updated the vcard search formulary, so it only mentions the usage of wildcard when using a storage that actually supports it (I think only Mnesia lacks it).

And I wonder if it's so important to support wildcard in vcard search for a Mnesia database...

**Comment by mremond** [comment]
I do not think wildcard search is that important in Mnesia, as Mnesia is used to target small user base.
